[Feature] 배송시작 이메일 전송

sangxxjin opened the pull request:
## 🌿 반영 브랜치
order/email -> main

## 🛠 작업 내용
- 이메일 전송 구현
- 메일 전송 실패 예외 처리
- 이메일 전송 테스트(mock 사용 x)
- 기존의 주문별로 상태 변경 -> 주문취합 후 이메일 별로 메일 전송, 주문 상태 변경

## 🔗 관련 이슈
Closes #30 

## 📂 추가 정보
아래의 코드 env파일 추가 ->  데이터는 슬랙에 남깁니다.
```
mail.username=
mail.password=
```

### 병합 후 크론잡 시간 변경하여 실제로 테스트 해야합니다.

## 💬 리뷰 요구사항 (선택)
- 메일 전송 실패 테스트 못했습니다.
- 이메일 전송까지만 테스트 하고 주문건 저장이 어려워 로직에 추가만 했습니다. -> DeliveryService쪽 꼼꼼하게 부탁드립니다.
- 테스트 코드 추가 할 부분이나 "이렇게 하면 더 좋지 않았을까?"와 같은 의견 자유롭게 공유해주세요.

sangxxjin committed:
chore: 이메일 전송 설정

sangxxjin committed:
feat: 이메일 전송 구현

sangxxjin committed:
test: 이메일 전송 테스트

본문:
Mock을 사용하지 않고 실제 전송

sangxxjin committed:
fix: 이메일 데이터 양식 수정

sangxxjin committed:
feat: EmailTemplateBuilder

본문: 클래스 분리 및 적용

sangxxjin committed:
Merge remote-tracking branch 'origin/main' into order/email

sangxxjin committed:
feat: 이메일 전송 로직 추가

본문:
기존에 주문별로 상태를 변경했는데 이메일 별로 주문 정보 수집 후 이메일 전송 및 주문 상태 변경

sangxxjin committed:
move: security 파일 경로 수정

sangxxjin committed:
feat: 이메일 전송 실패 예외처리

sangxxjin committed:
refactor: 가독성 향상

ChoiHyunSan commented on backend/src/main/java/com/ll/cafeservice/domain/delivery/service/DeliveryService.java:
```diff
@@ -3,29 +3,64 @@
+        return orderRepository.findByStatus(OrderStatus.WAITING);
+    }
+
+    private Map<String, List<Order>> groupOrdersByEmail(List<Order> orders) {
+        return orders.stream()
+            .collect(Collectors.groupingBy(Order::getEmail));
+    }
```
스트림으로 간결하게 정리해서 가독성이 정말 좋네요 :)

ChoiHyunSan commented on backend/src/main/java/com/ll/cafeservice/domain/delivery/service/DeliveryService.java:
```diff
@@ -3,29 +3,64 @@
+
+        Map<String, List<Order>> ordersGroupedByEmail = groupOrdersByEmail(orders);
+
+        for (Map.Entry<String, List<Order>> entry : ordersGroupedByEmail.entrySet()) {
+            String recipientEmail = entry.getKey();
+            List<String> orderNumbers = extractOrderNumbers(entry.getValue());
+            sendEmailToCustomer(recipientEmail, orderNumbers);
+            updateOrderStatusToCompleted(entry.getValue());
+        }
```
기존에 Complete 처리하는 로직과 같이 처리되고 있어서

Email단위로 주문번호 정보를 Send하고 완료처리 되는 것을 어렵지 않게 이해할 수 있었습니다 bb

ChoiHyunSan commented on backend/src/main/java/com/ll/cafeservice/global/email/EmailService.java:
```diff
@@ -0,0 +1,35 @@
+        sendHtmlEmail(recipientEmail, subject, text);
+    }
+
+    private void sendHtmlEmail(String to, String subject, String text) {
+        MimeMessage message = mailSender.createMimeMessage();
+        try {
+            MimeMessageHelper helper = new MimeMessageHelper(message, true);
+            helper.setTo(to);
+            helper.setSubject(subject);
+            helper.setText(text, true);
+            mailSender.send(message);
+        } catch (MessagingException e) {
+            throw new EmailServiceException("이메일 전송에 실패했습니다.", e);
+        }
+    }
+}
```
여기 조금 주의해야할 부분이 있어서 논의드리고 싶은 사항이 있습니다!

위의 Delivery 로직 상, 반복문을 돌며 
1) 이메일 보내기
2) Complete 처리

이 순서로 이루어지고 있는데, 만약 이메일이 잘못 쓰여진 경우 Throw 되면서 이메일은 보내지 못했지만 Complete는 처리되야 하지 않을까? 하는 로직적인 고민이 듭니다. 10개 중에 만약 2번째에서 이메일을 보내지 못한 경우에도 나머지 동작에 대해 처리되지 않는 게 아닐까? 하는 생각이 들었습니다.

혹시 이 부분은 어떻게 생각하시는지 궁금합니다!

sangxxjin replied:
@ChoiHyunSan  
pr날리고 검토하면서 생각했던 부분인데 예외처리가 아니라 로그를 남겨야 할 것 같습니다.
스케줄링을 통해 자동으로 이루어지는 부분인데 이메일 전송 실패를 예외처리 하는건 안되는 것 같고,
log를 남겨서 저희가 따로 확인할 수 있게 남겨야 할 것 같습니다.

다만, 이메일 전송같은 경우 잘못된 이메일을 입력했을 경우 실제로 이메일 전송이 되지 않을 뿐 에러가 발생하지 않습니다.

위의 제가 생각한 부분으로 적용하면 현산님이 걱정하셨던 부분 문제 없을 것 같습니다.

sangxxjin committed:
chore: hide mail info

ChoiHyunSan commented on backend/src/test/java/com/ll/cafeservice/global/EmailServiceTest.java:
테스트는 발송이 잘 되는지가 가장 중요하기 때문에, 우선 전송 테스트만 있으면 괜찮지 않을까 싶습니다!

ChoiHyunSan reviewed (commented):
전체적으로 코드 잘 봤습니다!

즉석으로 요구드린 사항인데 빠르게 잘 구현해주셔서 감사합니다.
가독성있게 작성해주셔서 리뷰에도 큰 어려움이 없었습니다.

리뷰드린 사항에서 논의드렸던 부분만 의견 적어주시면 좋을 것 같습니다 :)

junho3168 commented on backend/src/main/java/com/ll/cafeservice/global/email/EmailServiceException.java:
```diff
@@ -0,0 +1,11 @@
+package com.ll.cafeservice.global.email;
+
+public class EmailServiceException extends RuntimeException {
+    public EmailServiceException(String message) {
+        super(message);
+    }
+
+    public EmailServiceException(String message, Throwable cause) {
+        super(message, cause);
+    }
+}
```
EmailServiceException을 별도로 둬서 예외가 명확한 것 같습니다!

junho3168 commented on backend/src/main/java/com/ll/cafeservice/global/email/EmailTemplateBuilder.java:
```diff
@@ -0,0 +1,23 @@
+
+import java.util.List;
+
+public class EmailTemplateBuilder {
+
+    public static String buildOrderEmail(String recipientEmail, List<String> orderNumbers) {
+        StringBuilder orderList = new StringBuilder();
+        for (String orderNumber : orderNumbers) {
+            orderList.append("<li style='padding: 8px; border-bottom: 1px solid #eee;'>" + orderNumber + "</li>");
+        }
+
+        return "<html><body style='font-family: Arial, sans-serif; line-height: 1.6; background-color: #f4f4f9; padding: 20px;'>" +
+            "<div style='max-width: 600px; margin: 0 auto; background-color: #ffffff; padding: 20px; border-radius: 8px; box-shadow: 0 4px 6px rgba(0,0,0,0.1);'>" +
+            "<h2 style='color: #333;'>안녕하세요, " + recipientEmail + "님</h2>" +
+            "<p style='font-size: 16px; color: #555;'>주문 발송 메일을 전송드립니다. 아래는 귀하의 주문번호입니다:</p>" +
+            "<ul style='list-style-type: none; padding-left: 0;'>" + orderList.toString() + "</ul>" +
+            "<p style='font-size: 16px; color: #555;'>감사합니다!</p>" +
+            "<p style='font-size: 14px; color: #888;'>프로그래머스 백엔드 4회차 14팀</p>" +
+            "</div>" +
+            "</body></html>";
+    }
```
별도로 EmailTemplateBuilder를 두셔서 나중에 다시 쓰거나, 유지보수하기 편해보이네요!

junho3168 reviewed (commented):
작성하신 코드 잘 봤습니다 ㅎㅎ
작업 속도가 굉장히 빠르신 것 같습니다. 많이 배우고 갑니다! 

고생하셨습니다!!

sangxxjin committed:
fix: exception -> logger

본문:
스케줄링을 통해 메일을 전송함으로 exception에서 logger로 변경하였습니다.
log를 사용하지 않고 logger를 사용한 이유는 + 연산자를 사용하여 이메일 명을 남기기 위함입니다.
